### PR TITLE
Feature/validation engine

### PR DESCRIPTION
**Pull request by stanislaw-glogowski**
Connected #447 

The following pull request adds json schema validation to policy, condition and plugins params.
It also creates new `GET /schemas/:type?/:name?` endpoint in admin rest api.

**Example**:
`GET /schemas/policy`

**Response**:
```json
{
  "query": {
    "type": "policy"
  },
  "schemas": [{
    "name": "proxy",
    "type": "policy",
    "schema":{
      "$id": "http://express-gateway.io/schemas/proxy.json",
      "type": "object",
      "properties": {
        "serviceEndpoint": {"$ref": "config.json#/gateway/serviceEndpoint"},
        "changeOrigin": {"type": "boolean"},
        "strategy": {"type": "string", "enum": ["round-robin"]}
      },
      "required": ["serviceEndpoint"]
    }
  }, {
    "name": "oauth2",
    "type": "policy",
    "schema": {}
  }],
  ...
}
```

### CODE REVIEW

File: lib/policies/proxy/index.js [hunk 1]
```diff
@@ -1,3 +1,12 @@
 module.exports = {
-  policy: require('./proxy')
+  policy: require('./proxy'),
+  schema: {
+    type: 'object',
+    properties: {
+      serviceEndpoint: {type: 'string'},
```
Comment (DrMegavolt) [inline lib/policies/proxy/index.js]: I understand that this is work in progress, just a reminder, pls add all options from corresponding middleware

**Comment by DrMegavolt** [comment]
one small note:
properties like jscode assume some code editor (maybe not now, but in future)
and log/expression have kind of template syntax in some properties

most likely it will be helpful for UI to understand that these are not just strings
cc @kolbinski

**Comment by XVincentX** [comment]
I'll take over this PR.

File: lib/conditions/index.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
+const express = require('express');
+const logger = require('../logger').policy;
+const schemas = require('../schemas');
+const predefined = require('./predefined');
+const conditions = {};
+
+function register ({name, handler, schema}) {
+  let validate;
```
Comment (XVincentX) [inline lib/conditions/index.js]: ToDo: Change this function so it does not throw an exception. If it's returning a boolean, it means it's caller responsibility whether to throw an exception or not.

File: lib/conditions/index.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  try {
+    validate = schemas.register('condition', name, schema);
+  } catch (err) {
+    logger.warn(`warning: invalid schema for ${name} condition`, err);
+    return null;
+  }
+
+  conditions[name] = (req, config) => {
+    try {
+      validate(config);
```
Comment (XVincentX) [inline lib/conditions/index.js]: Note to self — again, do not throw an exception.

File: lib/conditions/predefined.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+module.exports = [
+  {
+    name: 'always',
+    handler: () => true,
+    schema: {}
+  }, {
+    // Not sure if anyone would ever use this in real life, but it is a
+    // "legitimate" condition, and is useful during tests.
+    name: 'never',
+    handler: (_req) => false,
```
Comment (XVincentX) [inline lib/conditions/predefined.js]: Nip: not required parameter.

File: lib/plugins.js [hunk 5]
```diff
@@ -30,6 +31,10 @@ module.exports.load = function ({config}) {
       if (semver.lt(engineVersion, plugin.version)) { // TODO: versioning of plugins
         logger.warn(plugin.version, 'is higher than engine version:', engineVersion, 'trying to load');
       }
+
+      // register schema and validate settings
+      schemas.register('plugin', pluginName, plugin.schema)(settings);
```
Comment (XVincentX) [inline lib/plugins.js]: Not handled exception.

File: lib/schemas/index.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  schemas: predefined
+});
+
+const buildKey = (type, name) => `${type}:${name}`;
+
+const registered = {};
+
+function register (type, name, schema) {
+  let compiled;
+  if (schema && typeof schema === 'object') {
```
Comment (XVincentX) [inline lib/schemas/index.js]: Worthless type check.

File: lib/schemas/index.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    compiled = ajv.compile(schema);
+  } else {
+    schema = {};
+    compiled = () => true;  // empty validator
+  }
+
+  registered[buildKey(type, name)] = {
```
Comment (XVincentX) [inline lib/schemas/index.js]: ajv has a built in caching mechanism. This can be removed.

**Comment by XVincentX** [comment]
@DrMegavolt I've modified couple of things and it should now be ready (with real passing tests this time), but I need another pairs of eyes.

I've also updated #422 in order to remember we should destroy `schemas` during the shutdown phase — but this is going to be another story.

File: lib/policies/index.js [hunk 8]
```diff
@@ -10,12 +12,32 @@ class Policies {
...
+
+    const validate = schemas.register('policy', name, schema);
+
+    policy.policy = (params, ...args) => {
+      const validationResult = validate(params);
+      if (validationResult.isValid) {
+        return action(params, ...args);
+      } else {
+        logger.warn(`warning: policy ${name} params validation failed`, validationResult.error);
+        return null;
```
Comment (DrMegavolt) [inline lib/policies/index.js]: Does this means that if I have incorrect schema I won't be able to use policy ? 
can it be  `return action(params, ...args);` here as well ?

File: lib/policies/key-auth/index.js [hunk 9]
```diff
@@ -1,5 +1,6 @@
-'use strict';
-
 module.exports = {
-  policy: require('./keyauth')
+  policy: require('./keyauth'),
+  schema: {
+    // TODO: add schema
```
Comment (DrMegavolt) [inline lib/policies/key-auth/index.js]: ##### Options Reference

* `apiKeyheader`:
  - 'Authorization'
  - the name of the header that should contain the API key
* `apiKeyHeaderScheme`:
  - 'apiKey'
  - the enforced scheme in the header
* `apiKeyHeader`:
  - 'Authorization'
  - name of the header that should contain api key
* `apiKeyHeaderScheme`:
  - 'apiKey'
  - enforce a schema in the header
* `disableHeaders`:
  - false
  - disable apikey lookup in headers
* `disableHeadersScheme`:
  - false
  - disable verification of scheme in header
* `apiKeyField`:
  - 'apiKey'
  - name of field to check in query parameter
* `disableQueryParam`:
  - false
  - set to true to disable api key lookup in query string

File: lib/policies/rate-limit/index.js [hunk 10]
```diff
@@ -1,5 +1,15 @@
-'use strict';
-
 module.exports = {
-  policy: require('./rate-limit')
+  policy: require('./rate-limit'),
+  schema: {
+    type: 'object',
+    properties: {
+      rateLimitBy: {type: 'string'},
+      windowMs: {type: 'integer'},
```
Comment (DrMegavolt) [inline lib/policies/rate-limit/index.js]: btw, can it be some description here or link to the doc? I'm ok with current version, just thinking how UI can show  help info

File: lib/schemas/index.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+function find (type = null, name = null) {
+  if (type && name) {
+    const item = ajv.getSchema(buildKey(type, name));
+    if (item) {
+      return [{ type, name, schema: item.schema }];
+    }
+  } else if (type) {
+    return registeredKeys
+      .filter(item => item.type === type)
```
Comment (DrMegavolt) [inline lib/schemas/index.js]: can we join L45 + L49 else case using condition: 
`.filter(item => !type || item.type === type)` ?

**Comment by DrMegavolt** [comment]
Some comments, but in general is almost good to go

File: lib/schemas/index.js (hunk 11 above)
Comment (XVincentX) [reply lib/schemas/index.js]: Good idea. Addressed in 0638256ff00fb402b653f9bac0238dae84fc2ec4

File: lib/policies/rate-limit/index.js (hunk 10 above)
Comment (XVincentX) [reply lib/policies/rate-limit/index.js]: Good point — JSON schema supports `title` and `description` property so we can definitely decorate those once we'll be able to forward this data to the UI.

File: lib/policies/index.js (hunk 8 above)
Comment (XVincentX) [reply lib/policies/index.js]: It depends — the idea behind the schema validation should be preventing the user from using the plugin if the provided parameters are not correct, therefore throwing an exception when this happens isn't it? Maybe @kevinswiber knows more about that?

**Pull request merged**